### PR TITLE
Improve management of Dart component instances

### PR DESCRIPTION
**Pull request by greglittlefield-wf**
___Note: for optimal experience when viewing the diff, [use the whitespace-ignoring flag: `?w=1`](https://github.com/cleandart/react-dart/pull/127/files?w=1).___

# Problem
Dart component instances were previously tied to the JS props and the Dart `ReactDartComponentInternal` objects.

 This led to confusing lifecycle of the Dart components, as well as issues when attempting to clean up leaks in that lifecycle (see https://github.com/cleandart/react-dart/pull/125).
> Unmounted component instances were retained by the original `ReactElement`'s props via the `ReactDartComponentInternal` object, which is used to tie the Dart component instance to the JS component instance.

# Changes
- Improve management of Dart component instances by storing them on the JS instances. This ensures that the lifecycle of Dart instances is in sync with the JS instances.
    - __Breaking change:__ Remove `ReactDartComponentInternal.component` - obviated by new JS instance storage
    - __Breaking change:__ Remove `ReactDartComponentInternal.isMounted` - not used, wasn't accurate after unmount

#### Other changes
- Add script to concatenate React and custom JS files sourced from `js_src` directory
    - Did this mostly for my own sanity to ensure I didn't mess anything up when copying custom JS from file to file
    - This resulted in a large number of whitespace changes, which can be ignored in this PR (see instructions at top about whitespace-agnostic diff viewing)
- Add transformer so tests can be run in DDC using `--pub-serve` flag

# Testing
- Run tests in content-shell, dart2js, and DDC locally to ensure that everything works as expected
- Verify that the memory leak is fixed:
    - Check out test branch: https://github.com/greglittlefield-wf/react-dart/tree/internal_instance_refactor-leak_test
    - Load `example/leak.html` with the dev tools open and following the instructions in the console
- Verify tests pass in over_react and WSD test branches:
    - https://github.com/Workiva/over_react/compare/master...greglittlefield-wf:internal_instance_refactor
    - https://github.com/Workiva/web_skin_dart/compare/master...greglittlefield-wf:internal_instance_refactor

# Areas of Regression
Component lifecycle

### CODE REVIEW

**Comment by greglittlefield-wf** [comment]
Please review: @aaronlademann-wf @jacehensley-wf @clairesarsam-wf @joelleibow-wf @trentgrover-wf

File: pubspec.yaml [hunk 1]
```diff
@@ -10,3 +10,7 @@ dependencies:
   js: '^0.6.0'
 dev_dependencies:
   test: '^0.12.18+1'
+
+transformers:
```
Comment (jacehensley-wf) [inline pubspec.yaml]: Do we need to include `dart2js` too? I was never sure of that

File: lib/react_client.dart [hunk 2]
```diff
@@ -194,50 +193,41 @@ final ReactDartInteropStatics _dartInteropStatics = (() {
...
   });
 
   /// Wrapper for [Component.componentWillMount].
-  void handleComponentWillMount(ReactDartComponentInternal internal) => zone.run(() {
-    internal.isMounted = true;
-    internal.component
-        ..componentWillMount()
-        ..transferComponentState();
+  void handleComponentWillMount(Component component) => zone.run(() {
+    component
```
Comment (jacehensley-wf) [inline lib/react_client.dart]: I'm surprised you used cascades here 😆

Comment (greglittlefield-wf) [inline lib/react_client.dart]: It's up to my discretion to use them 😜

File: pubspec.yaml (hunk 1 above)
Comment (greglittlefield-wf) [inline pubspec.yaml]: Nope, I believe that one is implied at the end if it isn't added explicitly

File: lib/react_client.dart [hunk 3]
```diff
@@ -194,50 +193,41 @@ final ReactDartInteropStatics _dartInteropStatics = (() {
...
+      ..initComponentInternal(internal.props, jsRedraw, getRef, jsThis)
+      ..initStateInternal();
 
-    internal.component = component;
-    internal.isMounted = false;
-    internal.props = component.props;
-
-    component.initStateInternal();
+    // Return the component so that the JS proxying component can store it,
+    // avoiding an interceptor lookup.
```
Comment (aaronlademann-wf) [inline lib/react_client.dart]: 🍻

**Comment by clairesarsam-wf** [comment]
+10

* Using the `internal_instance_refactor-leak_test` branch, component instances are not retained after unmount.
* `web_skin_dart` integration branch build succeeds ([view it here](https://ci.webfilings.com/build/984783))
* Tests in the `over_react` integration branch pass locally

**Pull request merged**